### PR TITLE
When to use synchronous - #145.

### PR DESCRIPTION
**Pull request by cynthia**
Needs more polish.

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/218.html" title="Last updated on Nov 3, 2020, 8:47 AM UTC (44502b7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/218/d0ad23a...44502b7.html" title="Last updated on Nov 3, 2020, 8:47 AM UTC (44502b7)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1218,6 +1218,20 @@ discuss.</p>
...
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when minting a new API.
+Providing an API in synchronous form simplifies the usage from the caller's perspective.
+Moreover, it reduces the amount of infrastructural setup overhead needed for the function call.
+
+The general rule of thumb for an API that should be synchronous is the following:
+
+  * The API call is not expected to fail.
```
Comment (annevk) [inline index.bs:1229]: Why does this matter?

Comment (cynthia) [reply index.bs:1229]: The synchronous part, or the not-failing part?

Comment (annevk) [reply index.bs:1229]: Not failing. A lot of non-promise returning methods have some kind of failure scenario.

Comment (torgo) [reply index.bs:1229]: Just revisiting this today. Is there a compromise text we could introduce here? @cynthia

**Comment by plinss** [comment]
Wondering if this could be better stated as reasons when to make methods async, and then say everything else should be sync.

File: index.bs (hunk 1 above)
Comment (cynthia) [reply index.bs:1229]: This was a suggestion from @kenchris - I think we can probably remove that clause and add it back when we have more context.

File: index.bs [hunk 2]
```diff
@@ -1218,6 +1218,20 @@ discuss.</p>
...
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when minting a new API.
+Providing an API in synchronous form simplifies the usage from the caller's perspective.
+Moreover, it reduces the amount of infrastructural setup overhead needed for the function call.
+
+The general rule of thumb for an API that should be synchronous is the following:
+
+  * The API call is not expected to be gated behind a permission.
```
Comment (kenchris) [inline index.bs:1229]: , now or in the future.

File: index.bs [hunk 3]
```diff
@@ -1218,6 +1218,20 @@ discuss.</p>
 </div>
 
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when minting a new API.
```
Comment (kenchris) [inline index.bs:1223]: adding/creating instead of minting

File: index.bs [hunk 4]
```diff
@@ -1218,6 +1218,20 @@ discuss.</p>
...
+Providing an API in synchronous form simplifies the usage from the caller's perspective.
+Moreover, it reduces the amount of infrastructural setup overhead needed for the function call.
+
+The general rule of thumb for an API that should be synchronous is the following:
+
+  * The API call is not expected to be gated behind a permission.
+  * The execution time is short and deterministic.
+
+No API call is completely fail-proof; the expectation above is for the general case.
```
Comment (kenchris) [inline index.bs:1232]: Remove

File: index.bs [hunk 5]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
 </div>
 
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when minting a new API.
+Providing an API in synchronous form simplifies the usage from the caller's perspective.
```
Comment (alice) [inline index.bs:1224]:
```suggestion
Synchronous APIs are simpler to use,
and need less infrastructure set-up (such as making functions `async`).
```

File: index.bs [hunk 6]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
...
+Moreover, it reduces the amount of infrastructural setup overhead needed for the function call.
```
Comment (alice) [inline index.bs:1225]:
```suggestion
```

File: index.bs [hunk 7]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
 </div>
 
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when minting a new API.
```
Comment (alice) [inline index.bs:1223]:
```suggestion
Synchronous APIs should be considered when creating a new API.
```

File: index.bs [hunk 8]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
 </div>
 
 
+<h3 id="synchronous">Use synchronous when appropriate</h3>
+
+Synchronous APIs should be considered when creating a new API.
```
Comment (alice) [inline index.bs:1223]:
```suggestion
Where possible, prefer synchronous APIs when designing a new API.
```

Switch to active voice.

File: index.bs [hunk 9]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
...
+Synchronous APIs are simpler to use,
+and need less infrastructure set-up (such as making functions `async`).
+
+The general rule of thumb for an API that should be synchronous is the following:
```
Comment (alice) [inline index.bs:1227]:
```suggestion
An API should generally be synchronous if the following rules of thumb apply:
```

File: index.bs [hunk 10]
```diff
@@ -1218,6 +1218,18 @@ discuss.</p>
...
+
+  * The API at no point is expected to be gated behind a permission or a dialog, such as a device selector.
```
Comment (alice) [inline index.bs:1229]:
```suggestion
  * The API is not expected to ever be gated behind a permission prompt, or another dialog such as a device selector.
```

**Review by alice — APPROVED** [review]
Advice looks good, just some editorial tweaks.

**Comment by annevk** [comment]
I would add "any kind of I/O" to the list of exceptions and maybe also "potential for the answer to be in a different process" given modern browser architectures.

**Comment by cynthia** [comment]
> I would add "any kind of I/O" to the list of exceptions and maybe also "potential for the answer to be in a different process" given modern browser architectures.

Reflected in latest commit. I dropped the "in a different process", because "accessing another process" will be done either through filesystem or network. (although the intent of that sentence including IPC might not be obvious to all spec authors, now that I think of it...)

**Comment by annevk** [comment]
Maybe add "including inter-process communication" or some such?

**Comment by cynthia** [comment]
> Maybe add "including inter-process communication" or some such?

**Comment by annevk** [comment]
"Including" acknowledges that while also reminding folks that might not realize, right? (You could make it a parenthetical.) However, I think the more important thing is to somehow point out that even if it's not the case today, it might be tomorrow. And perhaps not even cross-process, but simply running in a different thread.

**Comment by cynthia** [comment]
I threw in IPC as an example, for the thread example that would be a lock - so also added that.

**Review by torgo — APPROVED** [review]

**Review by alice — APPROVED** [review]

**Pull request merged**